### PR TITLE
fix AR(1) log likelihood

### LINKED ISSUE #2284: AR1 log likelihood
I think the [log likelihood](https://github.com/pymc-devs/pymc3/blob/255e8fcfb875a498f76c3148ceea135e781f61f9/pymc3/distributions/timeseries.py#L47) calculation for the AR1 model is wrong. 

```python
class AR1(distribution.Continuous):

    ...

    def logp(self, x):
        k = self.k
        tau_e = self.tau_e

        x_im1 = x[:-1]
        x_i = x[1:]
        boundary = Normal.dist(0., tau=tau_e).logp

        innov_like = Normal.dist(k * x_im1, tau=tau_e).logp(x_i)
        return boundary(x[0]) + tt.sum(innov_like) + boundary(x[-1])
```
The last term in the log likelihood should not be there.  

```python
import pymc3
from scipy.stats import norm

with pymc3.Model() as t:
    my_ar1 = pymc3.AR1('y',0,1,shape=2)

print('pymc3 log lik     :', t['y'].logp({'y':np.zeros(2,)}) )
print('scipy log lik(n=2):', norm.logpdf([0,0]).sum() )
print('scipy log lik(n=3):', norm.logpdf([0,0,0]).sum() )
```
yields
```
pymc3 log lik     : -2.756815599614018
scipy log lik(n=2): -1.83787706641
scipy log lik(n=3): -2.75681559961
``` 
I will try to write some tests following what's in test_distributions.py.  BTW, this is fantastic software -- thanks for all your efforts.

### PR DESCRIPTION
**Pull request by eph**
Fixes #2284 

I've never contributed to an open source project, so I'm not sure if a pull request if overkill here.

### CODE REVIEW

**Comment by junpenglao** [comment]
Welcome! No contribution is too small ;-)
Another point of reference is the implementation in `statsmodels`: [doc](http://www.statsmodels.org/0.6.1/generated/statsmodels.tsa.ar_model.AR.loglike.html#statsmodels.tsa.ar_model.AR.loglike); [code](http://www.statsmodels.org/0.6.1/_modules/statsmodels/tsa/ar_model.html#AR.loglike).

File: pymc3/distributions/timeseries.py [hunk 1]
```diff
@@ -44,7 +44,7 @@ def logp(self, x):
         boundary = Normal.dist(0., tau=tau_e).logp
 
         innov_like = Normal.dist(k * x_im1, tau=tau_e).logp(x_i)
-        return boundary(x[0]) + tt.sum(innov_like) + boundary(x[-1])
```
Comment (twiecki) [inline pymc3/distributions/timeseries.py:47]: Yeah, this looks like a bug to me.

File: pymc3/tests/test_distributions.py [hunk 2]
```diff
@@ -284,6 +284,10 @@ def mvt_logpdf(value, nu, Sigma, mu=0):
     return log_pdf
 
 
+def AR1_logpdf(value, k, tau_e):
+    return (sp.norm(loc=0,scale=1/np.sqrt(tau_e)).logpdf(value[0]) +
```
Comment (twiecki) [inline pymc3/tests/test_distributions.py]: pep8 wants spaces around math operators.

Comment (twiecki) [reply pymc3/tests/test_distributions.py]: But yeah, this test is a bit odd as it basically repcliates the likelihood. `statsmodels` is an idea but we don't want that as a dependency, but could be optional. @junpenglao?

Comment (junpenglao) [reply pymc3/tests/test_distributions.py]: I dont think we should use `statsmodels` in the test. I am thinking more rewriting our AR1 logp similar to the (exact) unconditional maximum likelihood as in `statsmodels`. We should take this opportunity to extend our AR1 model into AR(n).

Comment (eph) [reply pymc3/tests/test_distributions.py]: - Sorry about the test.  I didn't know the best way to write one without introducing additional dependencies.  
- As it written now, the AR(1) likelihood calculation is conditional on $x_0 = 0$, which seems a bit arbitrary to me.  It seems more natural to conditional on the initial observation.  
- You could use the unconditional distribution to calculate the likelihood of the initial observation (as in stats models), but this only makes sense if the process is stationary ($ |k| < 1 $).  Maybe something like `init` in `GaussianRandomWalk`?
- I am trying to write a more general AR RV, but I'm getting a bit tripped up by creating the matrix of lagged observations in theano.

**Comment by twiecki** [comment]
@eph Any progress?

**Comment by junpenglao** [comment]
@twiecki I think this is waiting for after 3.1rc4

**Comment by twiecki** [comment]
@junpenglao Definitely, but can start moving some things into the 3.2 branch when they are done.

**Comment by ferrine** [comment]
Is it ready?

**Comment by junpenglao** [comment]
I think we can merge this and improve the test later.

**Comment by ferrine** [comment]
Don't you think we'll forget about test. Or do we need test here?

**Comment by junpenglao** [comment]
@eph wrote a test here but it basically replicates the likelihood (see @twiecki's comment above).
But it actually wont be a big problem if we are going to rewrite the logp later closer to the form in `statsmodels`

**Comment by junpenglao** [comment]
pinging @eph

**Comment by eph** [comment]
I have written a more general (p lag) autoregressive model.  Should I replace AR1 with this class (after tests are written of course).  What do you want the tests written against?  I know statsmodels was mentioned, but 1) the loglikelihood calculation there is not completely clear to me and 2) I thought we didn't want to keep that dependency.  

```python 
class AR(distribution.Continuous):
    R"""
    Autoregressive process with p lags.

    .. math::

       x_t = \rho_0 + \rho_1 x_{t-1} + \ldots + \rho_p x_{t-p} + \epsilon_t,
       \epsilon_t \sim N(0,\sigma^2)

    
    The innovation can be parameterized either in terms of precision
    or standard deviation. The link between the two parametrizations is
    given by

    .. math::

       \tau = \dfrac{1}{\sigma^2}


    Parameters
    ----------
    rho : tensor
        Vector of autoregressive coefficients. 

    sd : float 
        Standard deviation of innovation (sd > 0).

    tau : float
        Precision of innovation (tau > 0).

    constant: bool (optional, default = False)
        Whether to include a constant. 
    """

    def __init__(self, rho, sd=None, tau=None,
                 constant=False, *args, **kwargs):
        
        super(AR, self).__init__(*args, **kwargs)
        tau, sd = get_tau_sd(tau=tau, sd=sd)
        self.sd = tt.as_tensor_variable(sd)
        self.tau = tt.as_tensor_variable(tau)

        self.mean = tt.as_tensor_variable(0.)

        rho = tt.as_tensor_variable(rho,ndim=1)
        if constant:
            self.p = rho.shape[0] - 1
        else:
            self.p = rho.shape[0]

        #if self.p <= 0:
        #    raise ValueError('too few lags.')

        self.constant = constant
        self.rho = rho
        

    def logp(self, value):

        y = value[self.p:]
        results,_ = scan(lambda l, obs, p: obs[p-l:-l],
                         outputs_info=None,sequences=[tt.arange(1,self.p+1)],
                         non_sequences=[value, self.p])
        x = tt.stack(results)

        if self.constant:
            y = y - self.rho[0]
            eps = y - self.rho[1:].dot(x)
        else:
            eps = y - self.rho.dot(x)

        innov_like = Normal.dist(mu=0.0, tau=self.tau).logp(eps)
        return tt.sum(innov_like)
```

**Comment by twiecki** [comment]
Yes, general AR would be great. Tests are difficult here, I wouldn't block on it. Perhaps just a regression test with fixed values? Or maybe a simple case where the result can be confirmed by hand?

**Comment by junpenglao** [comment]
There are conflicts @eph

File: pymc3/distributions/timeseries.py [hunk 3]
```diff
@@ -44,16 +45,93 @@ def logp(self, x):
...
+    .. math::
+
+       \tau = \dfrac{1}{\sigma^2}
+
+
+    Parameters
+    ----------
+    rho : tensor
+        Vector of autoregressive coefficients. 
+
```
Comment (ferrine) [inline pymc3/distributions/timeseries.py]: We usually do not use blank lines between

File: pymc3/distributions/distribution.py [hunk 4]
```diff
@@ -88,7 +88,7 @@ def getattr_value(self, val):
 
     def _repr_latex_(self, name=None, dist=None):
         return None
-
+
```
Comment (junpenglao) [inline pymc3/distributions/distribution.py]: This should no appear.

File: pymc3/distributions/distribution.py [hunk 5]
```diff
@@ -127,11 +127,6 @@ def __init__(self, shape=(), dtype=None, defaults=('mode',),
                 dtype = 'int64'
         if dtype != 'int16' and dtype != 'int64':
             raise TypeError('Discrete classes expect dtype to be int16 or int64.')
-
```
Comment (junpenglao) [inline pymc3/distributions/distribution.py]: This should not be deleted.

**Comment by junpenglao** [comment]
hi @eph, could you please edit the PR following the review? We should merge it after test pass.

**Comment by ferrine** [comment]
The notebook needs to be refined

**Comment by junpenglao** [comment]
Hi @eph, hope you dont mind I step in to fix the conflict and the notebook. 

As for this PR, how should we process? The test should be further refined, and the general AR logp could be improved as well (without using `theano.scan`). But in my opinion, the pm.timeseries needs some refactoring and further documentation, so it might be better to address these in a separate PR.

**Comment by junpenglao** [comment]
OK, after discuss with @twiecki i am merging this now. Further work on AR is referenced in #2491.

**Comment by junpenglao** [comment]
Thanks for the contribution @eph!

**Comment by eph** [comment]
Hi -- sorry for dropping the ball at the end of this PR.  Thanks for merging it in!  My hope is to implement a vector autoregression next.  How would build the lag matrix without looping  (`theano.scan`)?  Some kind of algebraic operation? or some other theano reshape / broadcasting procedure?

**Comment by junpenglao** [comment]
@eph no problem ;-)
I have no given much thought about how to build the logp without theano.scan, but judging from the implementation from statsmodel it seems possible? Just build a large (sparse) covariance matrix?
Maybe we can leverage some of our GP modules?
We can further discuss it in #2491.

**Pull request merged**